module/irrecoverable/example_test.go

```diff
@@ -0,0 +1,136 @@
+package irrecoverable_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/lifecycle"
+)
+
+type TemporaryError struct {
+	message string
+}
+
+func (s TemporaryError) Error() string {
+	return s.message
+}
+
+func Example() {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	starts := 0
+	var componentFactory module.ComponentFactory
+	componentFactory = func() (module.Component, error) {
+		var err error
+		starts++
+		if starts > 1 {
+			err = errors.New("Fatal! No restarts")
+		} else {
+			err = TemporaryError{"Restart me!"}
+		}
+
+		return NewExampleComponent(starts, err), nil
+	}
+
+	var onError module.OnError
+	onError = func(err error, triggerRestart func()) {
+		// check to make sure it's safe to restart the component
+		var tmp TemporaryError
+		if errors.As(err, &tmp) {
+			fmt.Printf("Restarting component after fatal error: %v\n", err)
+			triggerRestart()
+			return
+		}
+
+		// it's not safe to restart, shutdown instead
+		fmt.Printf("An unrecoverable error occurred: %v\n", err)
+		// shutdown other components. it might make sense to just panic here
+		// depending on the circumstances
+		cancel()
+	}
+
+	// run the component. this will block until onError fails to restart the component
+	// or its context is cancelled
+	err := module.RunComponent(ctx, componentFactory, onError)
+	if err != nil {
+		fmt.Printf("Error returned from RunComponent: %v\n", err)
+	}
```

Comment:
This is perhaps a good place to hint at 2 things (in comments):
- if the start of the component itself errors, you'll see an error here because the restart would not make sense,
- if the component itself needs to start components, you can:
    * reuse the same context, and 
    * implement an innerComponentFactory
    * implement an innerOnError
    and you'll have a hierarchical model, where the parent components supervise and restart their children.